src/config.ts

```diff
@@ -128,23 +130,93 @@ export class CacheConfig {
     self.workspaces = workspaces;
 
     let keyFiles = await globFiles("rust-toolchain\nrust-toolchain.toml");
+
+    hasher = crypto.createHash("sha1");
+
     for (const workspace of workspaces) {
       const root = workspace.root;
       keyFiles.push(
         ...(await globFiles(
-          `${root}/**/Cargo.toml\n${root}/**/Cargo.lock\n${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
+          `${root}/**/rust-toolchain\n${root}/**/rust-toolchain.toml`,
         )),
       );
+
+      // Get name of workspace crates.
+      const ws_crates = new Set(
+        (await workspace.getPackages()).map(p => p["name"])
```

Comment:
This here is a problem, as it internally calls `cargo metadata`.

We had to remove a call that happens in the "restore"/pre-phase of the action:
https://github.com/Swatinem/rust-cache/commit/3f2513fdf484da3bcfe60b1b40921055681bcdbf#diff-32b32ae92adc890fbd79b1d74f84cf0351380d3e0282f9a9f01352763140541dL46

This was because under some circumstances, probably because of a bug in cargo itself, `cargo metadata` will actually go ahead and compile a bunch of stuff. And by "compile a bunch of stuff", it did a full `cargo build`. Don’t ask me why because it is a mystery to me, but we experienced that problem on the https://github.com/getsentry/relay repo.

Long story short: `getPackages` is a big no-no in the pre-phase. (might be good to even add a big comment that function explaining this)

This is a bit unfortunate, but I believe the workaround for that is very simple:
Just check if you have a `path` dependency, and zero-out the version of that. The way cargo works, `path` dependencies are only a thing in local workspaces and always preferred over versions. Once you publish and someone else installs the package as a dependency, the `version` will be used instead.

Reply:
> Just check if you have a `path` dependency, and zero-out the version of that.

well, that wouldn’t work for the lockfile though :-(

Reply:
Actually… We are iterating all the `Cargo.toml` files, so we don’t need the `cargo metadata` invocation in that case because iterating over the `toml` files already gives us all the workspace dependency names. And we don’t have to care about all the target names either.

Reply:
Thanks!
And I just realized that you can rename dependencies in `Cargo.toml`:

```
serde2 = { package = "serde", version = "1" }
serde = { path = "...", version = "1" }
```

Looking at the `Cargo.lock`, I found that dependencies with `path` does not have `[[package]].source`, so we can figure that out all by looking at `Cargo.lock` or `Cargo.toml` independently!

Reply:
Also found that `[[package]].checksum` will not present for workspace dep, so I will detect them by looking at both `source` and `checksum`.

Reply:
Thanks, I've pushed the fix.